server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -298,7 +299,13 @@ public void innerOnResponse(Result result) {
                         @Override
                         public void onFailure(Exception t) {
                             try {
-                                onShardFailure(shardIndex, shard, shardIt, t);
+                                // It only happens when onPhaseDone() is called and executePhaseOnShard() fails hard with an exception.
```

Comment:
Ah, that would make sense, the failure of whole `executePhaseOnShard` was covered but not the failure in the listener, thank you @kkewwei